utilities/transactions/optimistic_transaction_test.cc

```diff
@@ -308,6 +311,205 @@ TEST_F(OptimisticTransactionTest, FlushTest2) {
+  s = txn->Commit();
+  // We should have checked two memtables
+  ASSERT_EQ(2, get_perf_context()->get_from_memtable_count);
+  // txn should fail because of conflict, even if the memtable
+  // has flushed, because it is still preserved in history.
+  ASSERT_TRUE(s.IsBusy());
+
+  get_perf_context()->Reset();
+  s = txn2->Commit();
+  // We should have checked two memtables
+  ASSERT_EQ(2, get_perf_context()->get_from_memtable_count);
+  ASSERT_TRUE(s.ok());
+
+  txn3->Put(Slice("foo2"), Slice("bar2"));
+  get_perf_context()->Reset();
+  s = txn3->Commit();
+  // txn3 is created after the active memtable is created, so that is the only
+  // memtable to check.
+  ASSERT_EQ(1, get_perf_context()->get_from_memtable_count);
+  ASSERT_TRUE(s.ok());
+
+  SetPerfLevel(PerfLevel::kDisable);
+
+  delete txn;
+  delete txn2;
+  delete txn3;
+}
+
+// Trigger the condition where some old memtables are skipped when doing
+// TransactionUtil::CheckKey(), and make sure the result is still correct.
+TEST_F(OptimisticTransactionTest, CheckKeySkipOldMemtable) {
+  const int kAttemptHistoryMemtable = 0;
+  const int kAttemptImmMemTable = 1;
+  for (int attempt = kAttemptHistoryMemtable; attempt <= kAttemptImmMemTable;
+       attempt++) {
+    options.max_write_buffer_number_to_maintain = 3;
+    Reopen();
+
+    WriteOptions write_options;
+    ReadOptions read_options;
+    ReadOptions snapshot_read_options;
+    ReadOptions snapshot_read_options2;
+    string value;
+    Status s;
+
+    txn_db->Put(write_options, Slice("foo"), Slice("bar"));
+    txn_db->Put(write_options, Slice("foo2"), Slice("bar"));
+
+    Transaction* txn = txn_db->BeginTransaction(write_options);
+    ASSERT_TRUE(txn);
+
+    Transaction* txn2 = txn_db->BeginTransaction(write_options);
+    ASSERT_TRUE(txn);
+
+    snapshot_read_options.snapshot = txn->GetSnapshot();
+    txn->GetForUpdate(snapshot_read_options, "foo", &value);
+    ASSERT_EQ(value, "bar");
+    txn->Put(Slice("foo"), Slice("bar2"));
+
+    snapshot_read_options2.snapshot = txn2->GetSnapshot();
+    txn2->GetForUpdate(snapshot_read_options2, "foo2", &value);
+    ASSERT_EQ(value, "bar");
+    txn2->Put(Slice("foo2"), Slice("bar2"));
+
+    // txn updates "foo" and tx2 updates "foo2", and now a write is
+    // issued for "foo", which conflicts with txn but not txn2
+    s = txn_db->Put(write_options, "foo", "bar");
+    ASSERT_OK(s);
+
+    if (attempt == kAttemptImmMemTable) {
+      // For the second attempt, hold the flush from happening
+      rocksdb::SyncPoint::GetInstance()->LoadDependency(
+          {{"OptimisticTransactionTest.CheckKeySkipOldMemtable",
+            "FlushJob::Start"}});
+      rocksdb::SyncPoint::GetInstance()->EnableProcessing();
+    }
+
+    // force a memtable flush. The memtable should still be kept
+    FlushOptions flush_ops;
+    if (attempt == kAttemptHistoryMemtable) {
+      txn_db->Flush(flush_ops);
+    } else {
+      assert(attempt == kAttemptImmMemTable);
```

Comment:
It is immediately clear to me why holding the flush make the things more complicated and worth the testing. Is not max_write_buffer_number_to_maintain=3 guarantees that the two memtbales will kept around with or without flush?

Reply:
It's just to cover the scenario. These are two different pieces of code, although the mechanism is the same. This is for the immutable memtable case: https://github.com/facebook/rocksdb/pull/4941/files#diff-d08b9e18884ca1e424184576672d16e6R3426

and this is the historic memtable case: https://github.com/facebook/rocksdb/pull/4941/files#diff-d08b9e18884ca1e424184576672d16e6R3426